vortex-array/src/array/bool/compute/filter.rs

```diff
@@ -1,67 +1,80 @@
+use crate::compute::{FilterFn, FilterIter, FilterMask};
 use crate::{ArrayData, IntoArrayData};
 
 impl FilterFn for BoolArray {
-    fn filter(&self, mask: &FilterMask) -> VortexResult<ArrayData> {
-        filter_select_bool(self, mask).map(|a| a.into_array())
+    fn filter(&self, mask: FilterMask) -> VortexResult<ArrayData> {
+        let validity = self.validity().filter(&mask)?;
+
+        let buffer = match mask.iter()? {
+            FilterIter::Indices(iter) => {
+                filter_indices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::LazyIndices(iter) => {
+                filter_indices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::Slices(iter) => {
+                filter_slices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+            FilterIter::LazySlices(iter) => {
+                filter_slices(self.boolean_buffer(), mask.true_count(), iter)
+            }
+        };
+
+        Ok(Self::try_new(buffer, validity)?.into_array())
     }
 }
 
-fn filter_select_bool(arr: &BoolArray, mask: &FilterMask) -> VortexResult<BoolArray> {
-    let validity = arr.validity().filter(mask)?;
-
-    let selection_count = mask.true_count();
-    let out = if selection_count * 2 > arr.len() {
-        filter_select_bool_by_slice(&arr.boolean_buffer(), mask, selection_count)
-    } else {
-        filter_select_bool_by_index(&arr.boolean_buffer(), mask, selection_count)
-    };
-    BoolArray::try_new(out?, validity)
-}
-
-fn filter_select_bool_by_slice(
-    values: &BooleanBuffer,
-    mask: &FilterMask,
-    selection_count: usize,
-) -> VortexResult<BooleanBuffer> {
-    let mut out_buf = BooleanBufferBuilder::new(selection_count);
-    mask.iter_slices()?.for_each(|(start, end)| {
-        out_buf.append_buffer(&values.slice(start, end - start));
-    });
-    Ok(out_buf.finish())
+fn filter_indices(
+    buffer: BooleanBuffer,
+    indices_len: usize,
+    mut indices: impl Iterator<Item = usize>,
+) -> BooleanBuffer {
+    let src = buffer.values().as_ptr();
+    let offset = buffer.offset();
+
+    BooleanBuffer::collect_bool(indices_len, |_idx| {
+        let idx = indices.next().vortex_expect("iterator length must match");
+        unsafe { bit_util::get_bit_raw(src, idx + offset) }
+    })
 }
 
-fn filter_select_bool_by_index(
-    values: &BooleanBuffer,
-    mask: &FilterMask,
-    selection_count: usize,
-) -> VortexResult<BooleanBuffer> {
-    let mut out_buf = BooleanBufferBuilder::new(selection_count);
-    mask.iter_indices()?
-        .for_each(|idx| out_buf.append(values.value(idx)));
-    Ok(out_buf.finish())
+fn filter_slices(
+    buffer: BooleanBuffer,
+    indices_len: usize,
+    slices: impl Iterator<Item = (usize, usize)>,
+) -> BooleanBuffer {
+    let src = buffer.values();
+    let offset = buffer.offset();
+
+    let mut builder = BooleanBufferBuilder::new(indices_len.div_ceil(8));
```

Comment:
this should just be indices_len, BooleanBuffer::new takes length in bits

write a better test somewhere

<img width="572" alt="Screenshot 2024-11-18 at 11 17 04" src="https://github.com/user-attachments/assets/2bf7dc1a-5c9f-4d17-9eed-68c61b2a9f93">

Reply:
I don't know what you want me to test... that we don't over-allocate memory that we never use? 

I took it from here: https://github.com/apache/arrow-rs/blob/b1f5c250ebb6c1252b4e7c51d15b8e77f4c361fa/arrow-select/src/filter.rs#L520

Seems like Arrow's filtering is also wrong

Reply:
https://github.com/apache/arrow-rs/issues/6750